scripts/filters.py

```diff
@@ -0,0 +1,150 @@
+
+    # 2. Bandpass filter
+    raw.filter(l_freq=1, h_freq=50)
+    print("Applied bandpass filter: 1-50 Hz")
+
+    # 3. Apply montage
+    ch_names = ['AF3', 'F7', 'F3', 'FC5', 'T7', 'P7', 'O1', 'O2', 'P8', 'T8', 'FC6', 'F4', 'F8', 'AF4']
+    raw.pick_channels(ch_names)
+    montage = mne.channels.make_standard_montage('standard_1020')
+    raw.set_montage(montage)
+    print("Montage applied")
+
+    # 4. Run PyPREP
+    prep_params = {
+        "ref_chs": "eeg",
+        "reref_chs": "eeg",
+        "line_freqs": np.arange(50, raw.info['sfreq'] / 2, 50),
+        "max_iterations": 8,
+    }
+    prep_pipeline = PrepPipeline(raw, prep_params, raw.get_montage())
+    prep_pipeline.fit()
+
+    print("Bad channels interpolated:", prep_pipeline.interpolated_channels)
+    print("Original bad channels:", prep_pipeline.noisy_channels_original["bad_all"])
+    print("Bad channels after interpolation:", prep_pipeline.still_noisy_channels)
+
+    raw = prep_pipeline.raw.copy()
+
+    # 5. Re-reference to the average
+    raw.set_eeg_reference('average', projection=True)
+    print("Re-referenced to the average")
+
+    # 6. ICA
+    ica = mne.preprocessing.ICA(n_components=0.99, method='fastica', random_state=42)
+    ica.fit(raw)
+    ica.plot_components(inst=raw)  # Optional: Plot components for manual inspection
+    ica.apply(raw)
+    print("ICA applied and artifacts removed")
+
+    # 7. Welch's PSD
+    f, psd = welch(raw.get_data(), fs=raw.info['sfreq'], nperseg=1024)
+
+    return {"cleaned_raw": raw, "frequencies": f, "psd": psd}
+
+# Function to apply preprocessing to multiple files
+def apply_to_files(folder_path, keyword=None):
+    """
+    Apply preprocessing to all EDF files in a folder.
+
+    Parameters:
+    - folder_path (str): Path to the folder containing EDF files.
+    - keyword (str, optional): Filter files by keyword.
+
+    Returns:
+        dict: Results of preprocessing for each file.
+    """
+    results = {}
+    for raw, file_path in load_edf_files(folder_path, keyword):
+        print(f"Preprocessing file: {file_path}")
+        result = preprocess_raw(raw)
+        results[file_path] = result
+    return results
+
+# Function to plot Welch's PSD
+def plot_psd(psd_data, freqs, title="Power Spectral Density"):
+    """
+    Plot the Power Spectral Density using Welch's method.
+
+    Parameters:
+    - psd_data (np.ndarray): PSD values.
+    - freqs (np.ndarray): Corresponding frequencies.
+    - title (str): Title for the plot.
+    """
+    plt.figure(figsize=(10, 6))
+    plt.semilogy(freqs, np.mean(psd_data, axis=0))
+    plt.xlabel('Frequency (Hz)')
+    plt.ylabel('Power Spectral Density (uV^2/Hz)')
+    plt.title(title)
+    plt.grid()
+    plt.show()
+
+# Main execution
+if __name__ == "__main__":
+    # Path to the folder with EDF files
+    folder_path = "Cualsea"  # Change to your folder path
+    keyword = "EPOC"  # Filter by keyword or set to None to load all files
```

Comment:
Using the following suggestion, you can pass the arguments as part of the command line. Default for keyword is `None`, but it could be `EPOC`.

```suggestion
if __name__ == "__main__":
    import sys

    # Path to the folder with EDF files
    if len(sys.argv) >= 2:
        folder_path = sys.argv[1]  # Change to your folder path
        keyword = sys.argv[2] # Filter by keyword or set to None to load all files
    elif len(sys.argv) == 1:
        folder_path = sys.argv[1]
        keyword = None
    else:
        print("""Pass arguments in the command line for instance:

    filters.py /path/to/my/folder EPOC

    """)
```

Reply:
"""
Script for generating topographic maps from EEG data.

Description:
- This script processes EEG data stored in a CSV file. It generates topographic maps for different frequency bands (Delta, Theta, Alpha, Beta, Gamma) for each subject, condition, and measure (e.g., M1, M2, etc.).
- The script uses the MNE library to handle EEG data and generates visualizations (topographic maps) for all specified frequency bands.
- The user can specify the folder path and keyword as command-line arguments to customize the input location.

Arguments (passed via command line):
1. folder_path: Path to the folder where CSV files are located (mandatory).
2. keyword: Keyword to filter specific files (optional, defaults to None).

Example usage:
    python filters.py /path/to/my/folder EPOC
    python filters.py /path/to/my/folder
"""